File: src/Java/algorithms/treeandgraph/ConvertSortedArrayToBinarySearchTree.java
package algorithms.treeandgraph; 

import java.util.*; 
import lib.TreeNode; 

public class ConvertSortedArrayToBinarySearchTree {
    /* Function: Give an sorted array, construct a BST from it. 
     * Solution: Recursive. 
     */ 
    
    public TreeNode sortedArrayToBST(int[] num) {
        if (num == null || num.length == 0) {
            return null; 
        }
        
        return builder(num, 0, num.length - 1); 
    }
    
    private TreeNode builder(int[] num, int st, int ed) {
        if (st > ed) {
            return null; 
        }
        
        int md = st + (ed - st) / 2; 
        TreeNode root = new TreeNode(num[md]); 
        root.left = builder(num, st, md - 1); 
        root.right = builder(num, md + 1, ed); 
        
        return root; 
    }
}

File: src/CPP/src/leetcode/ConvertSortedArrayToBinarySearchTree.cpp
/*
 * [Source] https://leetcode.com/problems/convert-sorted-array-to-binary-search-tree/
 * [Difficulty]: Medium
 * [Tag]: Tree
 * [Tag]: Depth-first Search
 */

#include <iostream>

using namespace std;

// [Solution]: Pick the middle element and use as root. Then construct the left and right subtree.
// [Corner Case]:
class Solution {
};

/* Java solution
https://github.com/ZhenyingZhu/ClassicAlgorithms/blob/master/src/algorithms/treeandgraph/ConvertSortedArrayToBinarySearchTree.java
 */

/* Java solution
public class Solution {
    public TreeNode sortedArrayToBST(int[] num) {
		return sortedArrayToBST(num, 0, num.length-1);
    }
	
	public TreeNode sortedArrayToBST(int[] num, int start, int end){
		if(start>end) return null;
		int mid=(start+end)/2;
		TreeNode root=new TreeNode(num[mid]);
		root.left=sortedArrayToBST(num, start, mid-1);
		root.right=sortedArrayToBST(num, mid+1, end);
		return root;
	}
}
 */

int main() {
    Solution sol;

    return 0;
}
